### ISSUE TITLE
Grafana: authentication: generic oauth role mapping is not working( version 9.5.13 )

### ISSUE THREAD

**Issue by suhasdaivajna** [grafana/grafana#78977]
### What happened?

Enabled generic_oauth for my grafana instance . User can login okay , however the role assignment is not working as per documentation.
User info Json Response:
{
    "email": "john_doe@mycompany.com",
    "email_verified": true,
    "family_name": "John",
    "given_name": "Doe",
    "groups": [
        {
            "id": "7dcbe5a6286d4ddb168656d3079334cace",
            "name": "SiteAdmin"
        },
        {
            "id": "13c71b28ed7143e28dd22b988cc3d9b4ed",
            "name": "SiteAdmin"
        },
        {
            "id": "120442475ddddcc441ba2760b7ebdc0f991",
            "name": "Developer"
        }
    ],
    "name": "John Doe",
    "preferred_username": "john_doe@mycompany.com",
    "sub": "john_doe@mycompany.com"
}

grafana.ini 

    [auth.generic_oauth]
    api_url = https://xxxx/oauth2/v1/userinfo
    auth_url = https://xxx/oauth2/v1/authorize
    client_id = xxxx
    client_secret = xxxx
    enabled = true
    name = Oracle IDCS
    role_attribute_path = contains(groups[*].name, 'SiteAdmin') && 'Admin' || contains(groups[*].name, 'SiteAdmin') && 'Editor' || 'Viewer'
    scopes = openid profile email groups
    token_url = https://xxxxx/oauth2/v1/token
    use_pkce = true
    use_refresh_token = true
    [log]
    mode = console

### What did you expect to happen?

Expected the user to get mapped to "Admin" role based on the example provided instead user get assigned the default role "Viewer"

### Did this work before?

N/A

### How do we reproduce it?

Oauth step as per documentation - https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/#configure-role-mapping

### Is the bug inside a dashboard panel?

No

### Environment (with versions)?

Grafana: 9.5.13
OS: Alpine Linux v3.18
Browser: Chrome Version 119.0.6045.199

### Grafana platform?

Kubernetes

**Comment by haemka** [comment]
I've got the same problem with v10.4.1 installed via RPM package on CentOS (but the issue was also present in v10.2.3 RPM).

Current workaround for me is to set my config to
```
role_attribute_path = "'Admin'"
```
instead of
```
role_attribute_path = "contains(roles[*], 'admin') && 'Admin' || contains(roles[*], 'editor') && 'Editor' || 'Viewer'"
```
restart Grafana, delete my session in Keycloak, delete my Cookies for Grafana, re-login, make any changes I need Admin permissions for, and revert the changes above afterwards. All that while hoping no other user will log in during this procedure.

**Comment by nickzelei** [comment]
For anyone coming here, I was able to get this to work with Keycloak and the docs may need to be updated.

I'm currently running Keycloak 23.x via helm chart 19.3.3.

The role_attribute_path detailed in the docs links to the wrong path.

This will vary on your setup, but I created the `admin, editor, viewer` roles as client roles.
This causes them to appear in the token under `resource_access.<client-id>`. In my case, the client-id is simply `grafana`.

```
role_attribute_path: contains(resource_access.grafana.roles[*], 'admin') && 'Admin' || contains(resource_access.grafana.roles[*], 'editor' && 'Editor') || 'Viewer'
```

However, the roles path does not by default show up in the ID token. The mappers must be configured to allow them to appear there.
This can be done by going to `Client Scopes -> Client scope details -> client roles` and enabling `Add to ID token`.
You may also change the token claim name on this screen which is by default set to `resource_access.${client_id}.roles` like I detailed above.

After this was configured, my role was correctly mapped!!

**Comment by bmassemin** [comment]
I agree, the documentation is incomplete regarding the mapping of the roles in the user info.
Here the minimal configuration to make it work with the actual documentation:
`role_attribute_path = contains(roles[*], 'admin') && 'Admin' || contains(roles[*], 'editor') && 'Editor' || 'Viewer'`

In your client configuration, go to `Client scopes` tab, and select the dedicated scope. Then click `Add mapper` and select `By configuration`, then `User Realm Role`.

![Image](https://github.com/user-attachments/assets/fc07594b-e8ad-42ce-936f-f70deb94fa3c)

Only `Add to userinfo` is required, no need to add the roles to the `ID token` or the `access token`.

**Comment by radubuciuceanu** [comment]
Hitting the same problem with Grafana v11.2.2. It was perfectly working with version v9.3.0. Here are my configs:

Grafana:

```
role_attribute_path: contains(realm_access.roles[*], 'inowattio_grafana_admin') && 'Admin' || contains(realm_access.roles[*], 'inowattio_grafana_editor') && 'Editor' || 'Viewer'
```

Keycloak:

![Image](https://github.com/user-attachments/assets/d492ad0e-9591-4716-8d97-d6c8bdfce289)


Have checked the logs, the received token has the roles but in the end, I get this message:

```
logger=oauth.generic_oauth t=2024-10-23T11:56:35.174786869Z level=debug msg="User info result" result="Id: f91473b3-986b-4e47-a5e1-ec484a27225e, Name: Radu Buciuceanu, Email: censored.censored@censored.com, Login: censored.censored@censored.com, Role: , Groups: [], OrgRoles: map[]"

```

Could not find any solution for that so for now, think we'll downgrade to v9.3.0 to solve that...

**Comment by gdecicco** [comment]
Hello,

additional detail regarding the first message. Which is the actual problem.
We discovered that only with some complex idp user info response grafana do not behave well.

This is a working example:
**IDP Response**:
```
{
  "aud": [
    "9xxxxxx"
  ],
  "auth_time": 1736427573,
  "email": "username@domain.tld",
  "email_verified": "true",
  "family_name": "Surname",
  "given_name": "Name",
  "groups": [
    "group1,
    "group2"
  ],
  "iat": 1736427574,
  "iss": "xxxx/",
  "name": "xxxxxx",
  "rat": 1736427569,
  "sub": "xxxxxx"
}
```
**Grafana config**
```
role_attribute_path: contains(groups[], 'group1') && 'Admin' || contains(groups[], 'group2') && 'Editor' || 'Viewer'
```

The user is correctly getting the Admin Role.


Whis this is a non working example:

**IDP Response** (The IDP is Oracle idcs) :
```
{
 "email": "username@domain.tld",
  "email_verified": true,
  "family_name": "Surname",
  "gender": "",
  "given_name": "Name",
  "groups": [
    {
      "id": "465d6ddf702e44a2ad0a1c2f6580b520",
      "name": "group1",
      "$ref": "http://xxxxx"
    },
    {
      "id": "36fc064c0aaf4c508e60aac2bf79450a",
      "name": "anothergroup",
      "$ref": "http://xxxx"
    }
  ],
  "name": "XXXXX",
  "sub": "xxxxxx"
}
```
**Grafana config**
```
role_attribute_path: contains(groups[*].name, 'group1') && 'Admin' || contains(groups[*].name, 'group2') && 'Editor' || 'Viewer'
```

The user is getting  getting the Viewer Role.


testing both the configuration on https://play.jmespath.org/ both should return "Admin" as role.

**Comment by max-pfeiffer** [comment]
> For anyone coming here, I was able to get this to work with Keycloak and the docs may need to be updated.
> 
> I'm currently running Keycloak 23.x via helm chart 19.3.3.
> 
> The role_attribute_path detailed in the docs links to the wrong path.
> 
> This will vary on your setup, but I created the `admin, editor, viewer` roles as client roles. This causes them to appear in the token under `resource_access.<client-id>`. In my case, the client-id is simply `grafana`.
> 
> ```
> role_attribute_path: contains(resource_access.grafana.roles[*], 'admin') && 'Admin' || contains(resource_access.grafana.roles[*], 'editor') && 'Editor' || 'Viewer'
> ```
> 
> However, the roles path does not by default show up in the ID token. The mappers must be configured to allow them to appear there. This can be done by going to `Client Scopes -> roles -> Mappers tab -> client roles` and enabling `Add to ID token`. You may also change the token claim name on this screen which is by default set to `resource_access.${client_id}.roles` like I detailed above.

I can confirm that this was the issue with my environment as well: Keycloak v26.1.0 and Grafana v11.4.0.

Per default Keycloak puts the roles into `resource_access` when you decode and check the claims in the JWT.

After adding it to the ID Token, the login with Grafana works.
